plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp

```diff
@@ -0,0 +1,189 @@
+         }
+      }
+
+      _snapshot_requests.emplace(producer_plugin::snapshot_schedule_information {{_snapshot_id++},{sri.block_spacing, sri.start_block_num, sri.end_block_num, sri.snapshot_description},{}});
+      x_serialize();
+   }
+
+   virtual void unschedule_snapshot(uint32_t sri) {
+      auto& snapshot_by_id = _snapshot_requests.get<by_snapshot_id>();
+      auto existing = snapshot_by_id.find(sri);
+      EOS_ASSERT(existing != snapshot_by_id.end(), chain::snapshot_request_not_found, "Snapshot request not found");
+      _snapshot_requests.erase(existing);
+      x_serialize();
+   }
+
+   virtual producer_plugin::get_snapshot_requests_result get_snapshot_requests() {
+      producer_plugin::get_snapshot_requests_result result;
+      auto& asvector = _snapshot_requests.get<as_vector>();
+      result.snapshot_requests.reserve(asvector.size());
+      result.snapshot_requests.insert(result.snapshot_requests.begin(), asvector.begin(), asvector.end());
+      return result;
+   }
+
+   // initialize with storage
+   void set_db_path(bfs::path db_path) {
+      _snapshot_db.set_path(std::move(db_path));
+      // init from db
+      if(fc::exists(_snapshot_db.get_json_path())) {
+         std::vector<producer_plugin::snapshot_schedule_information> sr;
+         _snapshot_db >> sr;
+         _snapshot_requests.insert(sr.begin(), sr.end());
+      }
+   }
+
+   // add pending snapshot info to inflight snapshot request
+   void add_pending_snapshot_info(const producer_plugin::snapshot_information & si) {
+      auto& snapshot_by_id = _snapshot_requests.get<by_snapshot_id>();
+      auto  snapshot_req   = snapshot_by_id.find(_inflight_sid);
+      if (snapshot_req != snapshot_by_id.end()) {
+         _snapshot_requests.modify(snapshot_req, [&si](auto& p) { 
+            if (!p.pending_snapshots) {
+               p.pending_snapshots = std::vector<producer_plugin::snapshot_information>();
+            }
+            p.pending_snapshots->emplace_back(si);       
+         });
+      }
+   }
+
+   // snapshot executor
+   void set_create_snapshot_fn(std::function<void(producer_plugin::next_function<producer_plugin::snapshot_information>)> fn) {
+      _create_snapshot = std::move(fn);
+   }
+
+   void execute_snapshot(uint32_t srid) {
+      _inflight_sid = srid;
+      auto next = [srid, this](const std::variant<fc::exception_ptr, producer_plugin::snapshot_information>& result) {
+         if(std::holds_alternative<fc::exception_ptr>(result)) {
+            try {
+               std::get<fc::exception_ptr>(result)->dynamic_rethrow_exception();
+            } catch(const fc::exception& e) {
+               elog( "snapshot creation error: ${details}", ("details",e.to_detail_string()) );
+               throw;
+            } catch(const std::exception& e) {
+               elog( "snapshot creation error: ${details}", ("details",e.what()) );
+               throw;
```

Comment:
Does this result in a graceful shutdown? Can you manually verify that it does.

Reply:
No, it was abort, changed this to app().quit() and now shutdown graceful. Thank you!!!